python/modules/BtagSFWeightQCDProducer.py

```diff
@@ -0,0 +1,144 @@
+            pt = jet.pt
+            eta = abs(jet.eta)
+            flavor = jet.hadronFlavour
+
+            if not ( pt > self.jetPtMin and eta < self.jetEtaMax): continue
+
+            if flavor == 5:
+                pt_bin = self.h_eff_b.GetXaxis().FindBin(pt); 
+                if pt_bin > self.h_eff_b.GetXaxis().GetNbins():
+                    pt_bin = self.h_eff_b.GetXaxis().GetNbins(); 
+                eta_bin = self.h_eff_b.GetYaxis().FindBin(eta); 
+                if eta_bin > self.h_eff_b.GetYaxis().GetNbins():
+                    eta_bin = self.h_eff_b.GetYaxis().GetNbins();
+
+                eff = self.h_eff_b.GetBinContent(pt_bin, eta_bin);
+
+            elif flavor == 4:
+                pt_bin = self.h_eff_c.GetXaxis().FindBin(pt); 
+                if pt_bin > self.h_eff_c.GetXaxis().GetNbins():
+                    pt_bin = self.h_eff_c.GetXaxis().GetNbins(); 
+                eta_bin = self.h_eff_c.GetYaxis().FindBin(eta); 
+                if eta_bin > self.h_eff_c.GetYaxis().GetNbins():
+                    eta_bin = self.h_eff_c.GetYaxis().GetNbins();
+
+                eff = self.h_eff_c.GetBinContent(pt_bin, eta_bin);
+
+            else:
+                pt_bin = self.h_eff_udsg.GetXaxis().FindBin(pt); 
+                if pt_bin > self.h_eff_udsg.GetXaxis().GetNbins():
+                    pt_bin = self.h_eff_udsg.GetXaxis().GetNbins(); 
+                eta_bin = self.h_eff_udsg.GetYaxis().FindBin(eta); 
+                if eta_bin > self.h_eff_udsg.GetYaxis().GetNbins():
+                    eta_bin = self.h_eff_udsg.GetYaxis().GetNbins();
+
+                eff = self.h_eff_udsg.GetBinContent(pt_bin, eta_bin);
+
+            if jet.btagDeepB > self.bDiscCut:
+                #check if eff is zero
+                if eff < 0.001:
+                    eff = 0.001
+
+		if flavor == 5:
+                	BTagWeightNHeavy_up   *= jet.btagSF_up * eff
+                	BTagWeightNHeavy_down *= jet.btagSF_down * eff
+		else:
+                	BTagWeightNLight_up   *= jet.btagSF_up * eff
+                	BTagWeightNLight_down *= jet.btagSF_down * eff
+
+                BTagWeightD      *= eff
+            else:
+                #check if eff is 1.0
+                if eff > 0.999:
+                    eff = 0.999
+
+		if flavor == 5:
+                	BTagWeightNHeavy_up   *= 1 - jet.btagSF_up * eff
+                	BTagWeightNHeavy_down *= 1 - jet.btagSF_down * eff
+		else:
+                	BTagWeightNLight_up   *= 1 - jet.btagSF_up * eff
+                	BTagWeightNLight_down *= 1 - jet.btagSF_down * eff
+
+                BTagWeightD      *= 1 - eff
+
+        self.out.fillBranch("BTagWeightHeavy_Up",   BTagWeightNHeavy_up / BTagWeightD)
+        self.out.fillBranch("BTagWeightHeavy_Down", BTagWeightNHeavy_down / BTagWeightD)
+        self.out.fillBranch("BTagWeightLight_Up",   BTagWeightNLight_up / BTagWeightD)
+        self.out.fillBranch("BTagWeightLight_Down", BTagWeightNLight_down / BTagWeightD)
```

Comment:
So this module is almost identical to the BtagSFWeightProducer.py, but storing the weight by flavour. Can we just add these variables to the original module. Do you need this for QCD smeared MC only, or all the MCs?

Reply:
Yes that would be great if we can add it to the original. we need it for qcd, data, ttbar, and wjets for each era. It might be easiest to just calculate it for all of them.